hashintel/hash
# Create `Kanban Board` block working with local data

## 🌟 What is the purpose of this PR?

This PR creates initial version of the Kanban block. It only works with local data.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1204000740778935/1204160628104410/f) _(internal)_
- [Design](https://www.figma.com/file/T10d2P2iNb1LdRC5a9J7vA/Blocks?node-id=1740-5423&t=rj4rYDVuiRCmIEEg-4) _(internal)_

## ⚠️ Known issues

- If `updateEntity` call takes a bit longer (e.g. while testing the block inside WordPress), and a consecutive update happens right after the `updateEntity` gets called, the update made by second debounced `updateEntity` call gets overrided by the previous one.  (check the video below to see how it looks)

*This issue can be reproduced by setting `simulateDatastoreLatency = {{ min: 500, max: 500  }}`* on `MockBlockDock`

https://user-images.githubusercontent.com/39553853/226212142-5b85e698-36e3-411c-9190-0fb7636975d6.mov


https://github.com/hashintel/hash/blob/6082cee5b413364b04bfb4bba067726ff32f86e9/blocks/kanban/src/components/board/board.tsx#L103-L115

*ℹ️ This code runs on each render to make sure `Board.tsx` respects changes of `blockEntity` props coming from `App.tsx`. (like [this example from official React docs](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes))*

And commenting out the code above prevents this issue. This issue probably happens by these steps

- `updateEntity` runs, and updates the `blockEntity`
- Because `blockEntity` is updated, the code above works
- But  in the meantime, user triggered another `updateEntity` call, and because `updateEntity` calls are debounced at `syncLocalStateToEntity` function, it runs with a delay, and probably works with stale data somehow 
  - This step definitely needs debugging. Maybe the [hacky](https://github.com/hashintel/hash/blob/6082cee5b413364b04bfb4bba067726ff32f86e9/blocks/kanban/src/components/board/board.tsx#L70-L101) `useState` usage at `syncLocalStateToEntity` causes this issue


## 🐾 Next steps

- [x] Fix known issue mentioned above (solution discussion on [Slack thread](https://hashintel.slack.com/archives/C02TWBTT3ED/p1679390791856139) _(internal)_)
- [x] Update block preview image (cc @nonparibus)
- [x] Write titles and descriptions for block `entity type` & `property types`
- [x] Update `README.md` (cc @nonparibus)

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  Start Kanban block in development mode
1.  Confirm that it works as expected, and looks like the design.

## 📹 Demo


https://user-images.githubusercontent.com/39553853/226211794-9475705e-3bd4-46c6-9344-5c4119c25850.mov
